### PR TITLE
Implement forget-by-default setting as a domain toggle.

### PR DESCRIPTION
**Pull request by goodov**
Resolves https://github.com/brave/brave-browser/issues/29723

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by goodov** [comment]
cc @pes10k for the correct texts we want here:

image

image

**Review by boocmp — APPROVED** [review]

**Review by pes10k — APPROVED** [review]
im not qualified to meaningfully review the implementation, but if the behavior matches what was discussed on slack, then terrific

**Comment by pes10k** [comment]
@goodov what happens to profiles with cookie content settings of "forget firstparty cookies"?

**Comment by goodov** [comment]
> @goodov what happens to profiles with cookie content settings of "forget firstparty cookies"?

As this is only in nightly and is not enabled globally, there's no any migration. You need to re-enable this for required domains.

**Comment by pes10k** [comment]
needing to re-enable is no problem and expected.  I was just wondering if my profile is going to end up in an undefined state bc of the short-lived origin-specific content settings, or if things will be fine

**Comment by goodov** [comment]
> needing to re-enable is no problem and expected. I was just wondering if my profile is going to end up in an undefined state bc of the short-lived origin-specific content settings, or if things will be fine

after your question I've added this to handle your situation:
```
   // Reset legacy host pattern.
  auto legacy_host_pattern = content_settings::CreateHostPattern(url);
  map->SetContentSettingCustomScope(
      legacy_host_pattern, ContentSettingsPattern::Wildcard(),
      ContentSettingsType::BRAVE_REMEMBER_1P_STORAGE, CONTENT_SETTING_DEFAULT);
```

will remove this bit in a next PR.

**Review by fmarier — APPROVED** [review]
The extra toggle is a good idea. It seems much less confusing then grouping this with the *block cookies* setting.

**Comment by goodov** [comment]
@brave/string-reviewers-team PTAL
@brave/chromium-src-reviewers PTAL

File: app/brave_settings_strings.grdp [hunk 1]
```diff
@@ -302,6 +299,9 @@
   <message name="IDS_SETTINGS_BRAVE_SHIELDS_DEFAULTS_ADVANCED_VIEW_LABEL" desc="Default Brave advanced view title label">
     Advanced view
   </message>
+  <message name="IDS_SETTINGS_FORGET_FIRST_PARTY_STORAGE" desc="Toggle label">
+    Forget me when I close a site
```
Comment (mkarolin) [inline app/brave_settings_strings.grdp:303]: @rmcfadden3 I think `Forget me when I close the site` [below](https://github.com/brave/brave-core/pull/18014/files?show-viewed-files=true&file-filters%5B%5D=#diff-f2536e2302c1f665f39278809f9413259fbcfcf5177441fbb21d3bcb51b48404R113) in `brave_shields_strings.grdp` in the context of the Shields panel is clear enough. But, do you think `Forget me when I close a site` in the context of the Shields settings on the settings page would be descriptive enough? Could this toggle benefit from having an explanation text added?

Comment (rmcfadden3) [reply app/brave_settings_strings.grdp:303]: @mkarolin — can you summarize here exactly what happens when this toggle is switched to "On" position for a site, and you then navigate away from that site? Like, does it clear all cookies for that site, and nuke it from your browser history?

Reason I ask is I think some people might already assume Brave does this, so we may need a slightly more descriptive text.

And, yes, to answer your question: I agree that adding explanation text would help.

Comment (mkarolin) [reply app/brave_settings_strings.grdp:303]: I think @goodov would be able to answer those questions better.

Comment (goodov) [reply app/brave_settings_strings.grdp:303]: @rmcfadden3 
The features clears cookies and any storage for the site (local storage, indexed db) after the website is closed and a short keep-alive period (30 seconds), but does not remove it from the history.

Comment (goodov) [reply app/brave_settings_strings.grdp:303]: ping @rmcfadden3 @mkarolin 
what's the decision? I can use `Forget me when I close the site` in both places for now (to not mess with a/the article in the shields popup and the settings page), but please tell me if it's okay.

Comment (rmcfadden3) [reply app/brave_settings_strings.grdp:303]: @pes10k @goodov @mkarolin — yeah, I suppose saying forget "me" is clear enough, in that it clears personal info, but not browsing history.

In the context of Shields settings on the settings page, I do think it would be helpful to include explanatory text. I'm open to your edits, but something like:

**`Forget me when I close this site`**
`Clears cookies, local storage, etc when you close a site.`

Comment (goodov) [reply app/brave_settings_strings.grdp]: @rmcfadden3 @pes10k 

Made it similar to "Clear browsing data" window:
image

**Review by mkarolin — APPROVED** [review]
LGTM
